### PR TITLE
net: tcp: Fix context leak during connect

### PR DESCRIPTION
**Pull request by neja20**
Replaced tcp_out() with tcp_out_ext() when sending SYN during tcp connect, so we can intercept error value, beacuse in such case the packet would not be sent at all and the stack would not trigger any mechanism to flush this context so it ends up leaking. These scenarios can arise when the underlaying interface is not properly configured or is disconnected. The conn is marked to be closed and is closed before exiting tcp_in().
Since we can now idefntify this case we can also exit early in the net_tcp_connect() function and not wait for any timeout.

### CODE REVIEW

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -2922,11 +2922,16 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 			verdict = NET_OK;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2922]: Typo in commit msg `s/idefntify/identify/`

**Review by jukkar — COMMENTED** [review]
LGTM, small nit about typo in commit msg

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -3918,7 +3923,10 @@ int net_tcp_connect(struct net_context *context,
 	(void)tcp_in(conn, NULL);
 
 	if (!IS_ENABLED(CONFIG_NET_TEST_PROTOCOL)) {
-		if ((K_TIMEOUT_EQ(timeout, K_NO_WAIT)) &&
+		if (conn->state == TCP_UNUSED || conn->state == TCP_CLOSED) {
+			ret = -errno;
+			goto out;
+		} else if ((K_TIMEOUT_EQ(timeout, K_NO_WAIT)) &&
 		    conn->state != TCP_ESTABLISHED) {
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:3930]: nit: Please fix the alignment here

**Review by rlubos — COMMENTED** [review]
Wow, that's a really good catch! Glad to see we have community on watch here, one more corner-case in our TCP stack covered.

Just please take a look at the nits reported and we're good.

File: subsys/net/ip/tcp.c (hunk 2 above)
Comment (neja20) [reply subsys/net/ip/tcp.c]: I hope I understood you correctly, you wanted the if statement aligned.

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**